packages/common/src/ContractInteractor.ts

```diff
@@ -940,10 +985,73 @@ calculateTransactionMaxPossibleGas: result: ${result}
     return await this.relayHubInstance.penalizer()
   }
 
+  private async _hubRelayRegistrarAddress (): Promise<Address> {
+    return await this.relayHubInstance.relayRegistrar()
+  }
+
   penalizerAddress (): Address {
     return this.penalizerInstance.address
   }
 
+  /**
+   * discover registered relays
+   * @param subset if set, then filter only to these relays
+   */
+  async getRegisteredRelays (subset?: string[], fromBlock?: number): Promise<RelayRegisteredEventInfo[]> {
+    const infoFromStorage = await this.getRegisteredRelaysFromRegistrar()
+    if (infoFromStorage != null) {
+      return infoFromStorage.filter(info => subset?.length === 0 || subset?.includes(info.relayManager))
+    } else {
+      return await this.getRegisteredRelaysFromEvents(subset, fromBlock)
+    }
+  }
+
+  async getRegisteredRelaysFromEvents (subsetManagers?: string[], fromBlock?: number): Promise<RelayRegisteredEventInfo[]> {
+    // each topic in getPastEvent is either a string or array-of-strings, to search for all.
+    const subsetManagersTopics = subsetManagers?.map(address2topic) as any as string
```

Comment:
like in your PR, each "topic"  it should be `string | string[]`
since I didn't change the type, I had to cast the returned array (see the "map") into a string. the getPAstEvent knows how to handle `string[]` for each indexed field.